runtime/lua/vim/func/_memoize.lua

```diff
@@ -33,31 +35,51 @@ local function resolve_hash(hash)
   return hash
 end
 
+--- @param weak boolean?
+--- @return table
+local create_cache = function(weak)
+  return setmetatable({}, {
+    __mode = weak ~= false and 'kv',
+  })
+end
+
 --- @generic F: function
 --- @param hash integer|string|fun(...): any
 --- @param fn F
---- @param strong? boolean
+--- @param weak? boolean
 --- @return F
-return function(hash, fn, strong)
+return function(hash, fn, weak)
   vim.validate({
     hash = { hash, { 'number', 'string', 'function' } },
     fn = { fn, 'function' },
+    weak = { weak, 'boolean', true },
   })
 
-  ---@type table<any,table<any,any>>
-  local cache = {}
-  if not strong then
-    setmetatable(cache, { __mode = 'kv' })
-  end
-
-  hash = resolve_hash(hash)
+  --- @type MemoObj
+  local obj = {
+    _cache = create_cache(weak),
+    _hash = resolve_hash(hash),
+    _weak = weak,
+    --- @param self MemoObj
+    invalidate = function(self, ...)
```

Comment:
Any objection to naming this `clear()` (see `:help dev-name-common`)? 

We may eventually need to introduce `dispose()` as a conventional name, but that probably doesn't fit here.

Reply:
I was also thinking `clear` or `reset`.